### DISCUSSION TITLE
Non density-based sensitivity analysis using adjoint solver

### DISCUSSION THREAD

**Discussion by alexthunderk** [NanoComp/meep#2002] [Q&A]
Hi meep community,

Just recently, I started using meep and it’s a great tool. Thanks for developing and maintaining it.
I came across the adjoint optimization and wanted to use it for a sensitivity analysis of design parameters, e.g., period, fill factor etc. of a grating coupler. Is that possible in meep? I found examples only for topology optimization using a material grid and haven’t been able to adapt it so far.
I created a class with an update-method so that it can be used as the design variables. However, the optimizer doesn’t update the parameters. I guess that’s because the derivative w.r.t. the parameters is not defined. Is that correct and is there a way to adapt the adjoint optimizer for this use case?

**Comment by lucasgrjn** [comment]
Hi,

I am actually interested in the same thing as you. Indeed, I need to reproduce the results of this article https://opg.optica.org/prj/fulltext.cfm?uri=prj-8-4-528&id=429433 (Inverse design of digital nanophotonic devices using the adjoint method) but unfortunately, until now, I have not been able to... (BTW, if I succeed, it can make a beautiful example).

If you succeed, please, share your solution !

A possible way of resolution for your issue will be to use the library autograd or jax. It will allow your simulation to include a gradient.

**Comment by smartalecH** [comment]
> Is that correct and is there a way to adapt the adjoint optimizer for this use case?

Sure, this is possible. But there are a few caveats concerning how you implement this within the current framework.

Note that to obtain sensitivity information w.r.t. the parameters you're describing, you need to use a levelset. Technically, the adjoint solver was meant to work with density-based methods (which are starkly different than levelsets). But we recently added some features that hybridize the two frameworks, provided the following conditions are met:

1. The parameters of interest are "mapped" to an underlying grid (essentially a density-grid). For example, your "mapping function" could simply look like

```python
import autograd.numpy as npa
x = npa.linspace(0,L,N) # where N is the number of MaterialGrid DOF, and L is grating length
def fx(Λ):
    return npa.sin(2*npa.pi/Λ*x) # where Λ is grating period
```
where the output of `fx` corresponds to the `MaterialGrid` design variables.

2. This function mapping must be "smooth" (you can't "threshold" or "project" yet).
3. the `MaterialGrid`'s `beta` parameter is set to `beta=np.inf`. In other words, the projection happens "under the hood".

The adjoint solver will provide the gradient, `∂F/∂fx` (where `F` is your objective function). You can then backpropagate the gradient through the mapping function to obtain the final sensitivity `∂F/∂Λ` (or any other additional parameters you may want).

**Comment by stevengj** [comment]
In particular, for a given shape (e.g. a polygon), if you define the material grid to be 0.5 + a signed distance function, and then set β=infinity, then the subpixel smoothing and the gradients will work.

**Comment by alexthunderk** [comment]
Hi,

Thanks a lot for your answers!
I’m still stuck on this topic and I don’t really know who to implement that properly.
I attached the code how I simulate the grating coupler currently (just looping over the parameters). If the simulation is correct, that would be a good tool to simulate regular grating couplers (In the attached file, I got a strange error while plotting the simulation setup, but the simulation seems to work).
However, I would like to change and optimize the gaps and grating lines individually, which is why I thought that the adjoint method would be beneficial. I don’t know if that would work with a function like you proposed,  @smartalecH.
If the optimization of the gaps, grating lines, fill factor and BOX thickness works, it would make a nice example, as @Dj1312 said, and a great tool.

[FGC_2D_fiber.txt](https://github.com/NanoComp/meep/files/8315158/FGC_2D_fiber.txt)

**Reply by smartalecH** [reply]
> However, I would like to change and optimize the gaps and grating lines individually

Take a look at #2011 (idea 3). Yes, it's possible to do this too. But the constraints listed [above](https://github.com/NanoComp/meep/discussions/2002#discussioncomment-2361602) must still be met.

**Reply by alexthunderk** [reply]
Thanks again!
Looking forward to the tutorial

**Comment by wllshr123** [comment]
Hello! I'm a novice user of meep. You said you have  examples only for topology optimization using a material grid. Can you share it with me? Thank you very much! Or do you have other optimization examples of vertical coupler with meep? 
Thanks a lot!
I'm looking forward to your reply!

**Reply by smartalecH** [reply]
Please keep questions about another discussion (#2393) within that discussion. Similarly, please don't spam other discussions in hopes of increasing engagement.
